src/Client/ConnectionPool.cpp

```diff
@@ -0,0 +1,81 @@
+#include <Client/ConnectionPool.h>
+
+#include <boost/functional/hash.hpp>
+
+namespace DB
+{
+
+ConnectionPoolPtr ConnectionPoolFactory::get(
+    unsigned max_connections,
+    String host,
+    UInt16 port,
+    String default_database,
+    String user,
+    String password,
+    String cluster,
+    String cluster_secret,
+    String client_name,
+    Protocol::Compression compression,
+    Protocol::Secure secure,
+    Int64 priority)
+{
+    Key key{
+        max_connections, host, port, default_database, user, password, cluster, cluster_secret, client_name, compression, secure, priority};
+
+    std::unique_lock lock(mutex);
+    auto it = pool.find(key);
+    ConnectionPoolPtr ret;
+    if (it != pool.end())
+        ret = it->second.lock();
+
+    if (!ret)
+    {
+        ret = std::make_shared<ConnectionPool>(
+            max_connections,
+            host,
+            port,
+            default_database,
+            user,
+            password,
+            cluster,
+            cluster_secret,
+            client_name,
+            compression,
+            secure,
+            priority);
+        if (it == pool.end())
```

Comment:
Seems that dangling `ConnectionPoolWeakPtr` will be left in the `pool` since you don't use custom deleter that will clean it up (not sure that it is a problem here though, but looks like it is better to avoid such things)

Reply:
> not sure that it is a problem here though, but looks like it is better to avoid such things

Sure.